File: apps/03-food-simple/backend/python/src/services/vector_search/postgres_service.py
import os
from typing import List, Dict, Any, Union
import asyncpg
from .base_vector_search import BaseVectorSearch

class PostgresService(BaseVectorSearch):
    """PostgreSQL vector search implementation"""
    
    def __init__(self):
        self.pool = None
        self.default_table = "recipes"
    
    async def connect(self):
        """Connect to PostgreSQL"""
        try:
            if self.pool is None:
                self.pool = await asyncpg.create_pool(
                    host=os.getenv("POSTGRES_HOST", "host.docker.internal"),
                    port=int(os.getenv("POSTGRES_PORT", 5432)),
                    database=os.getenv("POSTGRES_DB", "vectordb"),
                    user=os.getenv("POSTGRES_USER", "postgres"),
                    password=os.getenv("POSTGRES_PASSWORD", "postgres")
                )
        except Exception as e:
            raise Exception(f"Error connecting to PostgreSQL: {str(e)}")
    
    async def search_by_vector(self, vector: List[float], collection: str, limit: int = 5, return_content: bool = False) -> List[Dict[str, Any]]:
        """Search for similar vectors in the collection"""
        raise NotImplementedError("Method not supported")
    
    async def search_by_text(self, text: str, collection: str = None, limit: int = 5,
                           return_content: bool = False) -> Union[List[str], List[Dict[str, Any]]]:
        """Search for similar text in the collection"""
        try:
            # Connect if not already connected
            await self.connect()
            
            # Use the provided collection or default table
            table = collection if collection else self.default_table
            
            async with self.pool.acquire() as conn:
                query = f"""SELECT
                chunk,
                embedding <=> ai.ollama_embed('nomic-embed-text', $1) AS distance
                FROM public.{table}_embedding_store
                ORDER BY distance
                LIMIT $2;
                """
                
                rows = await conn.fetch(query, text, limit)
                
                # Format results based on return_content flag
                if return_content:
                    return [
                        {
                            "filename": "",
                            "content": row["chunk"]
                        }
                        for row in rows
                    ]
                else:
                    return [row["filename"] for row in rows]
        except Exception as e:
            raise Exception(f"Error searching in PostgreSQL: {str(e)}")
    
    async def generate_answer(self, question: str, collection: str) -> str:
        """Generate an answer using the vector search provider"""
        raise NotImplementedError("Method not supported")
    
    def get_provider(self) -> str:
        """Get the name of the vector search provider"""
        return "postgres"

File: apps/03-food-simple/backend/nodejs/src/services/vector-search/PostgresService.js
import pg from 'pg';
import VectorSearchInterface from './VectorSearchInterface.js';

class PostgresService extends VectorSearchInterface {
  constructor() {
    super();
    this.client = new pg.Client({
      host: process.env.POSTGRES_HOST || 'host.docker.internal',
      port: process.env.POSTGRES_PORT || 5432,
      database: process.env.POSTGRES_DB || 'vectordb',
      user: process.env.POSTGRES_USER || 'postgres',
      password: process.env.POSTGRES_PASSWORD || 'postgres'
    });
    this.defaultTable = 'recipes';
    this.connect();
  }

  async connect() {
    try {
      await this.client.connect();
    } catch (error) {
      console.error('Error connecting to PostgreSQL:', error);
      throw new Error(`Error connecting to PostgreSQL: ${error.message}`);
    }
  }

  async searchByText(text, collection, limit = 5, returnContent = false) {
    try {
      const query = `SELECT
      chunk,
      embedding <=> ai.ollama_embed('nomic-embed-text', $1) AS distance
      FROM public.${this.defaultTable}_embedding_store
      ORDER BY distance
      LIMIT $2;`
      const result = await this.client.query(query, [text, limit]);
      
      if (returnContent) {
        return result.rows.map(row => ({
          filename: "",
          content: row.chunk
        }));
      } else {
        return result.rows.map(row => row.filename);
      }
    } catch (error) {
      console.error(`Error searching by text in PostgreSQL table ${this.defaultTable}:`, error);
      throw new Error(`Error searching by text in PostgreSQL: ${error.message}`);
    }
  }

  getProvider() {
    return 'postgres';
  }
}

export default PostgresService; 